AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Subscribe/DataStoreObserveQueryOperation.swift

```diff
@@ -0,0 +1,375 @@
+///     - Generate first snapshot based on the internal state
+///     When the operation receives item changs
+///     - Batch them into batches of up to 1000 items or when 2 seconds have elapsed (`.collect(2s,1000)`)`
+///     - Update internal state of items based on the changed items
+///     - Generate new snapshot based on lates state of the items, with the items changed dedupped.
+@available(iOS 13.0, *)
+public class AWSDataStoreObserveQueryOperation<M: Model>: AsynchronousOperation, DataStoreObserveQueryOperation {
+
+    private let serialQueue = DispatchQueue(label: "com.amazonaws.AWSDataStoreObseverQueryOperation.serialQueue",
+                                            target: DispatchQueue.global())
+    private let itemsChangedPeriodicPublishTimeInSeconds: DispatchQueue.SchedulerTimeType.Stride = 2
+    private let itemsChangedMaxSize = 1_000
+
+    let modelType: M.Type
+    let modelSchema: ModelSchema
+    let predicate: QueryPredicate?
+    let sortInput: [QuerySortDescriptor]?
+
+    let storageEngine: StorageEngineBehavior
+    var dataStorePublisher: ModelSubcriptionBehavior
+
+    let observeQueryStarted: AtomicValue<Bool>
+    let isSynced: AtomicValue<Bool>
+    var currentItemsMap: [Model.Identifier: M]
+    var itemsChangedSink: AnyCancellable?
+    var dataStoreEventSink: AnyCancellable?
+
+    /// Internal publisher for `ObserveQueryPublisher` to pass events back to subscribers
+    let passthroughPublisher: PassthroughSubject<DataStoreQuerySnapshot<M>, DataStoreError>
+
+    /// External subscribers subscribe to this publisher
+    private let observeQueryPublisher: ObserveQueryPublisher<M>
+    public var publisher: AnyPublisher<DataStoreQuerySnapshot<M>, DataStoreError> {
+        return observeQueryPublisher.eraseToAnyPublisher()
+    }
+
+    init(modelType: M.Type,
+         modelSchema: ModelSchema,
+         predicate: QueryPredicate?,
+         sortInput: [QuerySortDescriptor]?,
+         storageEngine: StorageEngineBehavior,
+         dataStorePublisher: ModelSubcriptionBehavior) {
+        self.modelType = modelType
+        self.modelSchema = modelSchema
+        self.predicate = predicate
+        self.sortInput = sortInput
+        self.storageEngine = storageEngine
+        self.dataStorePublisher = dataStorePublisher
+
+        self.observeQueryStarted = AtomicValue(initialValue: false)
+        self.isSynced = AtomicValue(initialValue: false)
+        self.currentItemsMap = [:]
+        self.passthroughPublisher = PassthroughSubject<DataStoreQuerySnapshot<M>, DataStoreError>()
+        self.observeQueryPublisher = ObserveQueryPublisher()
+        super.init()
+        observeQueryPublisher.configure(operation: self)
+    }
+
+    override public func main() {
+        startObserveQuery()
+    }
+
+    override public func cancel() {
+        if let itemsChangedSink = itemsChangedSink {
+            itemsChangedSink.cancel()
+        }
+        passthroughPublisher.send(completion: .finished)
```

Comment:
i am wondering if `.finished` is correct to send in case of when operation is cancelled. do we need to differentiate between cancel and success events?

Reply:
i think a successfully cancellation action should result in a successful finished event for the subscriber since this action can be taken by the developer